keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,479 @@
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-allow-webhook-config>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: NoOpinion
+```
+Validation will allow multiple authorizers of type "Webhook" to be added to the config, but one authorizer each for other types. The ordering of this chain will be decided by the order specified in the file.
+
+The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and `subjectAccessReviewVersion` version accept values corresponding to flags `--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`, `--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version` respectively.
+
+This diagram shows how the configuration flags map to configuration file keys for a webhook:
+![](https://i.imgur.com/ZEuqcKe.jpg)
+
+The `onError` will be a required field which allow users to specify whether or not request is denied if the webhook errors out or is unreachable. This allows `SubjectAccessReviews` to be more definitive.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding flag is not supplied. While configuring authorization modes using the file config, the version supported by a webhook has to be mentioned using a required field `subjectAccessReviewVersion`.
+
+Authentication of webhooks Webhooks will continue to be configured by passing a path to a file resembling a `kubeconfig`. Eventually, we can make a decision about moving webhook configuration to the proposed authorization configuration.
+
+The config file would accept a path to the policy file if the auth mode type is ABAC.
+```
+#...
+authorizers:
+- type: ABAC
+  policyFile: <path to policy file>
+#...
+```
+
+The code path for enabling the above will only be triggered if the feature flag will be enabled until the time the feature flag is removed and configuring authorizer through a file becomes GA.
+
+> TODO: More details on the implementation will be charted once the summary, motivation, goals, non-goal and proposal are approved.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests would be added along with any new code introduced.
+
+Existing test coverage:
+- cmd/kube-apiserver/app/options: 65
+- staging/src/k8s.io/apiserver/pkg/authorization: NA
+- staging/src/k8s.io/apiserver/apis/config/v1: 54.5
+
+> Will add more packages to the list during implementation.
+
+##### Integration tests
+
+Integration tests would be added to ensure the following:
+- Authorization of requests work in the existing flag based mode (feature flag turned off)
+- Authorization of requests work with an apiserver bootstrapped with authorization configuration file (feature flag turned on)
```

Comment:
you need to check the failure mode handling too.

Reply:
Ack! We will amend the KEP.